### PR TITLE
End of Year: add stories Data Source and gestures

### PR DESCRIPTION
**Pull request by leandroalonso**
| 📘 Project: #376 |
|:---:|

This PR adds a data source protocol to populate the Stories view with stories and also adds a few gestures.

https://user-images.githubusercontent.com/7040243/195642253-8da709b6-d2da-43ef-8bdb-743f727fbe45.mp4

## About the data source

I tried to be as generic as possible with how we display stories to allow different cases to be used without much effort.

In order to show stories, we need to pass a data source that conforms to `StoriesDataSource`. Each story is simply a `View`.

This makes it possible for a story to be a custom view, an image, or a video. It's also generic enough to be used for other purposes in the future.

Take a look at the `TestStoriesDataSource` for an example.

The overall API is still in its early days, so we'll change it as the feature gets developed.

## To test

1. Enable the `endOfYear` flag in `FeatureFlag.swift`
2. Run the app
3. When the prompt appears, tap "View My 2022"
4. ✅ Check that each story is shown for 5s and that the story indicator (the rectangles at the top) are reflecting the state
5. ✅ Tap and hold, check that the current story is "paused". The rectangle stop progressing and the story doesn't change
6. ✅ While seeing the yellow story, tap the left portion of the screen, you should go back to the purple one
7. ✅ While seeing the purple story, tap the right portion of the screen, you should go to the next story (yellow)
8. ✅ Swipe down and make sure the view is dismissed ¶

¶ Swipe down can be improved with better effects (like with the view following the gesture) but this will be done in another task

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

File: podcasts/End of Year/StoriesModel.swift [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+class StoriesModel: ObservableObject {
+    @Published var progress: Double
+
+    @Published var currentStory: Int = 0
+
+    private let dataSource: StoriesDataSource
+    private let publisher: Timer.TimerPublisher
+    private var cancellable: Cancellable?
+    private var interval: TimeInterval = 5.seconds
```
Comment (emilylaguna) [inline podcasts/End of Year/StoriesModel.swift:12]: I wonder if in the future we should move this into the datasource to allow different story views to have different durations?

File: podcasts/End of Year/StoriesModel.swift [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    func story(index: Int) -> AnyView {
+        dataSource.storyView(for: index)
+    }
+
+    func next() {
+        progress = Double(Int(progress) + 1)
+    }
+
+    func previous() {
+        progress = Double(Int(progress) - 1)
```
Comment (emilylaguna) [inline podcasts/End of Year/StoriesModel.swift:50]: If I tap the previous action while on the first view the progress gets set to -1 which causes progress to be delayed by 5 seconds before it starts ticking up again.

https://user-images.githubusercontent.com/793774/195682972-c83ff059-76a0-49b7-8b83-0e1652ea9ecb.mov

File: podcasts/End of Year/StoriesModel.swift [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    func start() {
+        cancellable = publisher.autoconnect().sink(receiveValue: { _ in
+            var newProgress = self.progress + (0.01 / self.interval)
+
+            let currentStory = Int(newProgress)
+
+            if currentStory >= self.numberOfStories {
+                newProgress = 0
+                self.currentStory = 0
```
Comment (emilylaguna) [inline podcasts/End of Year/StoriesModel.swift]: What do you think about offloading this behavior to the datasource to allow it to decide how it wants to handle when we reach the end of the stories?

File: podcasts/End of Year/Views/StoriesView.swift [hunk 4]
```diff
@@ -63,36 +71,41 @@ struct StoriesView: View {
...
+                .onChanged { _ in
+                    model.pause()
+                }
+                .onEnded { value in
+                    let velocity = CGSize(
+                        width: value.predictedEndLocation.x - value.location.x,
+                        height: value.predictedEndLocation.y - value.location.y
+                    )
+
+                    // If a quick swipe down is performed, dismiss the view
```
Comment (emilylaguna) [inline podcasts/End of Year/Views/StoriesView.swift:101]: Are we also planning on adding a swipe next/back gesture?

**Review by emilylaguna — CHANGES_REQUESTED** [review]
Nice, this is shaping up! I found a bug with tapping back while on the first story (see video below). 

I also left a few other comments below. Let me know what you think.

File: podcasts/End of Year/StoriesModel.swift (hunk 1 above)
Comment (leandroalonso) [reply podcasts/End of Year/StoriesModel.swift:12]: Yes, that's the idea! I'll do it in another PR. :)

File: podcasts/End of Year/StoriesModel.swift (hunk 2 above)
Comment (leandroalonso) [reply podcasts/End of Year/StoriesModel.swift:50]: Good catch! Fixed in e44ffee77f93bb6d011212d0a7cd20f623607c65

File: podcasts/End of Year/StoriesModel.swift (hunk 3 above)
Comment (leandroalonso) [reply podcasts/End of Year/StoriesModel.swift]: This is also another idea I have for subsequent tasks. :)

File: podcasts/End of Year/Views/StoriesView.swift (hunk 4 above)
Comment (leandroalonso) [reply podcasts/End of Year/Views/StoriesView.swift:101]: Hmm, is swipe next/back gesture common in stories inside the same "context"?

I know it's a thing when you want to "advance" (to the next user, for example), but not for the next story.

Comment (emilylaguna) [reply podcasts/End of Year/Views/StoriesView.swift:101]: Good point. I don't think it's needed anymore.

**Review by emilylaguna — APPROVED** [review]
Nice work!

**Pull request merged**